src/approaches/grammar_search_invention_approach.py

```diff
@@ -772,6 +792,45 @@ def _evaluate_atom_trajectory(
         return score
 
 
+@dataclass(frozen=True, eq=False, repr=False)
+class _HeuristicCountBasedScoreFunction(_HeuristicBasedScoreFunction):  # pylint:disable=abstract-method
+    """Implement _evaluate_atom_trajectory() by using the induced operators to
+    compute estimated costs-to-go.
+
+    Then for each transition in the atoms_sequence, check whether the
+    transition is optimal with respect to the estimated costs-to-go. If
+    the transition is optimal and the sequence is not a demo, that's
+    assumed to be bad; if the transition is not optimal and the sequence
+    is a demo, that's also assumed to be bad.
+    """
+
+    def _evaluate_atom_trajectory(self, atoms_sequence: List[Set[GroundAtom]],
+                                  heuristic_fn: Callable[[Set[GroundAtom]],
+                                                         float],
+                                  ground_ops: Set[_GroundSTRIPSOperator],
+                                  is_demo: bool) -> float:
+        score = 0.0
+        for i in range(len(atoms_sequence) - 1):
+            atoms, next_atoms = atoms_sequence[i], atoms_sequence[i + 1]
+            best_h = float("inf")
+            on_sequence_h = float("inf")
+            for predicted_next_atoms in utils.get_successors_from_ground_ops(
+                    atoms, ground_ops, unique=False):
+                # Compute the heuristic for the successor atoms.
+                h = heuristic_fn(predicted_next_atoms)
+                best_h = min(h, best_h)
+                if predicted_next_atoms == next_atoms:
+                    assert on_sequence_h in [h, float("inf")]
+                    on_sequence_h = h
+            # Bad case 1: transition is optimal and sequence is not a demo.
+            if on_sequence_h == best_h and not is_demo:
+                score += CFG.grammar_search_off_demo_count_penalty
+            # Bad case 2: transition is not optimal and sequence is a demo.
+            elif on_sequence_h > best_h and is_demo:
+                score += CFG.grammar_search_on_demo_count_penalty
```

Comment:
checking myself before i go further: i think on_penalty should be higher than off_penalty

Reply:
correct hooray